torch/testing/_internal/common_distributed.py

```diff
@@ -877,6 +888,66 @@ def _check_return_codes(self, elapsed_time) -> None:
     def is_master(self) -> bool:
         return self.rank == 0
 
+# Utility base class for DDP based Multi Process Test cases
+# This abstracts the PG creation and deletion, the backends are selected based
+# on device type. The tests functions can be instantiated per device type using
+# common_device_type.instantiate_device_type_tests
+# other backends can add entry in backend() function
+class DistributedTestBase(MultiProcessTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self._spawn_processes()
+
+    def tearDown(self):
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    def backend(self, device) -> str:
+        if "cuda" in device:
+            return "nccl"
+        elif "hpu" in device :   # intel gaudi
+            return "hccl"
+        else :
+            return "gloo"
+
+    def device_handle(self, device) :
+        device_name = device.split(":")[0]
+        if hasattr(torch, device_name):
+            return getattr(torch, device_name, None)
+        else:
+            raise RuntimeError(
+                f"Device {device} library not found"
+            )
+
+    def create_pg(self, device):
+        num_visible_devices = self.device_handle(device).device_count()
+        store = torch.distributed.FileStore(self.file_name, num_visible_devices)
+        torch.distributed.init_process_group(
+            backend=self.backend(device),
+            world_size=self.world_size,
+            rank=self.rank,
+            store=store
+        )
+        if "nccl" in self.backend(device):
+            torch.cuda.set_device(self.rank)
+        return torch.distributed.distributed_c10d._get_default_group()
+
```

Comment:
Maybe we can leave this to child class?

Reply:
The rational behind adding this here was to allow for cleaner and shorter code in the child class. As it is meant to be derived and used it can be overwritten by the child class wherever required.

I would love to hear your thoughts on this though.

Reply:
There are two reasons:
1. Some TestClass may want to init distributed only once for the entire test suite. (to save time).
2. Some TestClass may want to pass different init options to `init_process_group`

Reply:
#### Regarding init options for TestClasses:

The current implementation allows child classes to overwrite the method, providing flexibility for different init options. However, if you think it would be beneficial, we could consider passing init options as parameters to the function which could perhaps make this easier on future users. 

#### Concerning single initialization of the distributed environment:
I understand the concern about initializing the distributed environment only once for the entire test suite.

To accommodate this, we could explore creating a separate class, modeled after `MultiProcContinuousTest`, which would allow for a single process group creation for the entire file. Alternatively, we could integrate this functionality within the existing class structure to allow for a more uniform implementation across test files.

As the current changes align with the existing implementation of `MultiProcessTestCase` while enabling easier device-agnostic implementation and they do not interfere with classes that initialize the process group only once, I believe it makes sense to limit the scope of this PR as it is. 

I'm more than willing to work on a follow-up PR to better accommodate these functionalities. This will allow other developers to benefit from these changes while I work on adding these functionalities. Your input on this would be valuable.


All in all having these functionalities in the base class enables easier usage of these test classes in independent test files without significantly reducing flexibility for the reasons mentioned above and hence I feel they should remain as they are and not be left to the child class.

I'm open to further discussion on these points and any other concerns you may have.